src/monitor/group_state_machine.c

```diff
@@ -295,14 +341,533 @@ ProceedGroupState(AutoFailoverNode *activeNode)
+	 * state, and now we need to select one of the failover candidates.
+	 *
+	 * The selection is based on candidatePriority. If the candidate with the
+	 * higher priority doesn't have the most recent LSN, we have it fetch the
+	 * missing WAL bits from one of the standby which did receive them.
+	 */
+	if (reportedLSNCount == candidateCount)
+	{
+		/* build the list of failover candidate nodes, ordered by priority */
+		List *candidateNodesGroupList =
+			GroupListCandidates(standbyNodesGroupList);
+
+		/* find the standby node that has the most advanced LSN */
+		AutoFailoverNode *mostAdvancedNode =
+			FindMostAdvancedStandby(standbyNodesGroupList);
+
+		AutoFailoverNode *selectedNode = NULL;
+
+		/*
+		 * Select the node to be promoted: we can pick any candidate with the
+		 * max priority, so we pick the one with the most recent LSN among
+		 * those of maxPriority.
+		 */
+		foreach(nodeCell, candidateNodesGroupList)
+		{
+			candidateNode = (AutoFailoverNode *) lfirst(nodeCell);
+
+			if (IsHealthy(candidateNode)
+				&& WalDifferenceWithin(candidateNode,
+									   primaryNode,
+									   PromoteXlogThreshold))
+			{
+				int cPriority = candidateNode->candidatePriority;
+				XLogRecPtr cLSN = candidateNode->reportedLSN;
+
+				if (selectedNode == NULL)
+				{
+					selectedNode = candidateNode;
+				}
+				else if (cPriority == selectedNode->candidatePriority
+						 && cLSN > selectedNode->reportedLSN)
+				{
+					selectedNode = candidateNode;
+				}
+				else if (cPriority < selectedNode->candidatePriority)
+				{
+					/*
+					 * Short circuit the loop, as we scan in decreasing
+					 * priority order.
+					 */
+					break;
+				}
+			}
+		}
+
+		/*
+		 * We might have selected a node to fail over to: start the failover.
+		 */
+		if (selectedNode != NULL)
+		{
+			char message[BUFSIZE];
+
+			/* do we have to fetch some missing WAL? */
+			if (selectedNode->reportedLSN == mostAdvancedNode->reportedLSN)
+			{
+				LogAndNotifyMessage(
+					message, BUFSIZE,
+					"Setting goal state of %s:%d to prepare_promotion "
+					"after %s:%d became unhealthy.",
+					selectedNode->nodeName, selectedNode->nodePort,
+					primaryNode->nodeName, primaryNode->nodePort);
+
+				AssignGoalState(selectedNode,
+								REPLICATION_STATE_PREPARE_PROMOTION,
+								message);
+
+				/* leave the other nodes in ReportLSN state for now */
+				return true;
+			}
+
+			/* so the candidate does not have the most recent WAL */
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to wait_forward "
+				"and goal state of %s:%d to wait_cascade "
+				"after %s:%d became unhealthy.",
+				selectedNode->nodeName, selectedNode->nodePort,
+				mostAdvancedNode->nodeName, mostAdvancedNode->nodePort,
+				primaryNode->nodeName, primaryNode->nodePort);
+
+			AssignGoalState(selectedNode,
+							REPLICATION_STATE_WAIT_FORWARD, message);
+
+			AssignGoalState(mostAdvancedNode,
+							REPLICATION_STATE_WAIT_CASCADE, message);
```

Comment:
What's the reason we don't put the primary node into draining here?

Reply:
In that branch the primary has been put into draining already...